horizomer/utils/tests/test_tree.py

```diff
@@ -457,6 +458,128 @@ def test_compare_branch_lengths(self):
         tree11 = TreeNode.read(['((a:1,(x:1,c:1)d:1)e:1,f:1)g:1;'])
         self.assertFalse(compare_branch_lengths(tree11, tree1))
 
+    def test_walk_copy(self):
+        tree1 = TreeNode.read(['(((a:1.0,b:0.8)c:2.4,(d:0.8,e:0.6)f:1.2)g:0.4,'
+                               '(h:0.5,i:0.7)j:1.8)k;'])
+        assign_supports(tree1)
+
+        # test pos = root
+        msg = 'Cannot walk from root of an rooted tree.'
+        with self.assertRaisesRegex(ValueError, msg):
+            walk_copy(tree1.find('k'), tree1.find('j'))
+        msg = 'Source and node are not neighbors.'
+
+        # test pos = derived
+        with self.assertRaisesRegex(ValueError, msg):
+            walk_copy(tree1.find('a'), tree1.find('b'))
+        with self.assertRaisesRegex(ValueError, msg):
+            walk_copy(tree1.find('c'), tree1.find('f'))
+        with self.assertRaisesRegex(ValueError, msg):
+            walk_copy(tree1.find('f'), tree1.find('j'))
+        with self.assertRaisesRegex(ValueError, msg):
+            walk_copy(tree1.find('f'), tree1.find('k'))
+
+        # test pos = basal
+        with self.assertRaisesRegex(ValueError, msg):
+            walk_copy(tree1.find('g'), tree1.find('a'))
+        with self.assertRaisesRegex(ValueError, msg):
+            walk_copy(tree1.find('g'), tree1.find('k'))
+
+        # pos = derived, move = up
+        res = TreeNode.read(['(a:1.0,((d:0.8,e:0.6)f:1.2,'
+                             '(h:0.5,i:0.7)j:2.2)g:2.4)c:0.8;'])
+        output = walk_copy(tree1.find('c'), tree1.find('a'))
+        for n1, n2 in zip(res.postorder(), output.postorder()):
```

Comment:
I think this comparison can be singled out as an independent function. So that we don't have to duplicate code.

Reply:
Is it better to add a comparison method in tree.py or test_tree.py?

Reply:
@nujinuji Should be in `tree.py`.